# A Python Quick Start Tutorial
# Lesson \#1: Basic data types and functions
## by Peter Mackenzie-Helnwein
University of Washington, Seattle, WA

pmackenz@uw.edu          
https://www.ce.washington.edu/facultyfinder/peter-mackenzie-helnwein

## Resources

   1. Python Docs: https://docs.python.org/3/
   
   1. Python Tutorial (comprehensive): https://docs.python.org/3/tutorial/index.html
   
   1. Python Library Reference (the nitty-gritty details): https://docs.python.org/3/library/index.html
   
   1. Everything else: http://google.com
   

## Discussing questions from the self-study assignment

This section is to discuss your questions


### Exercise 1

**Theory**:
Stress transformation

$$
\begin{aligned}
\sigma_{x}' &= \sigma_{x} \cos^2\theta + \sigma_{y} \sin^2\theta + 2\tau_{xy} \sin\theta \cos\theta \\
\sigma_{y}' &= \sigma_{x} \sin^2\theta + \sigma_{y} \cos^2\theta - 2\tau_{xy} \sin\theta \cos\theta \\
\tau_{xy}'  &= (\sigma_{y} - \sigma_{x}) \sin\theta \cos\theta + \tau_{xy} (\cos^2\theta - \sin^2\theta) \\
\end{aligned}
$$

**Given**:
Stress state:

$$
\sigma_{x} = 12~ksi~,
~~~
\sigma_{y} = -5.5~ksi~,
~~~
\tau_{xy} = 3.5~ksi 
$$

**Find**:
Components of stress in a rotated coordinate system for $\theta=25^\circ$.

In [2]:
# theta needs to be in radians
# sin and cos are not part of phython

from math import sin, cos, pi, radians

In [3]:
radians(90)

1.5707963267948966

In [44]:
sigma_x = 12
sigma_y = -5.5
tau_xy  = 3.5
theta   = radians(25.)

sx = sigma_x*cos(theta)**2 + sigma_y*sin(theta)**2 + 2*tau_xy*sin(theta)*cos(theta)
sy = sigma_x*sin(theta)**2 + sigma_y*cos(theta)**2 - 2*tau_xy*sin(theta)*cos(theta)
txy = (sigma_y-sigma_x)*sin(theta)*cos(theta)+ tau_xy*(cos(theta)**2 - sin(theta)**2)


print(sx, sy, txy) 

11.555547135673642 -5.055547135673642 -4.453132243388171


In [45]:
print("sigma_prime_x =" "{:8.4f}".format(sx));
print("sigma_prime_y =" "{:8.4f}".format(sy));
print("tau_prime_x   =" "{:8.4f}".format(txy));

sigma_prime_x = 11.5555
sigma_prime_y = -5.0555
tau_prime_x   = -4.4531


### Exercise 2

Convert the calculation from exercise 1 to a function

In [48]:
def stressTransform(sigx, sigy, tauxy, theta):
    # YOUR CODE
    th  = radians(theta)
    sx  = sigx * cos(th)**2 + sigy * sin(th)**2 + 2 *tauxy * sin(th) *cos(th)
    sy  = sigx * sin(th)**2 + sigy * cos(th)**2 - 2 *tauxy * sin(th) *cos(th)
    txy = (sigy-sigx) *sin(th) * cos(th) + tauxy*((cos(th)**2 - sin(th)**2))
    
    return (sx,sy,txy)

Test your function for 

* $\theta=0$ (what should the result look like?)
* $\theta=25^\circ$
* $\theta=180^\circ$

In [51]:
stress = stressTransform(12, -5.5, 3.5, 180)


In [55]:
print("sig_x  = {:8.4f} ksi \nsig_y  = {:8.4f} ksi \ntau_xy = {:8.4f} ksi ". format(*stress))

sig_x  =  12.0000 ksi 
sig_y  =  -5.5000 ksi 
tau_xy =   3.5000 ksi 


### Exercise 3

Let's use this function to study stress transformation.  In order to do so,

1. Create a list of 81 evenly spaced $\theta$-values from $0^\circ$ to $180^\circ$.
2. Compute the transformed stress for each of these values and collect the results in three lists: 

       sigmax_values = ...
       sigmay_values = ...
       tauxy_values = ...

3. print a nice table with 4 columns: $\theta$, $\sigma_x'$, $\sigma_y'$, $\tau_{xy}'$

In [72]:
sigmax_values = []
sigmay_values = []
tauxy_values  = []

# YOUR CODE HERE
print("     theta         sigma_x       sigma_y        tau_xy ")
print("======================================================")

lower  = 0;
upper  = 180;
length = 81;

theta = [lower + x*(upper-lower)/length for x in range(length)]
theta.append(180)

for i in theta:
    th = i;
    stress = stressTransform(12, -5.5, 3.5, th)
    sx     = stress[0]
    sy     = stress[1]
    txy    = stress[2]
    
    sigmax_values.append(sx)
    sigmay_values.append(sy)
    tauxy_values.append(txy)
    
    print("{:12.6f}, {:12.6f}, {:12.6f}, {:12.6f},". format(th, sx, sy, txy))
    
      

     theta         sigma_x       sigma_y        tau_xy 
    0.000000,    12.000000,    -5.500000,     3.500000,
    2.222222,    12.244912,    -5.744912,     2.811417,
    4.444444,    12.435727,    -5.935727,     2.105925,
    6.666667,    12.571298,    -6.071298,     1.387768,
    8.888889,    12.650810,    -6.150810,     0.661265,
   11.111111,    12.673784,    -6.173784,    -0.069215,
   13.333333,    12.640083,    -6.140083,    -0.799279,
   15.555556,    12.549908,    -6.049908,    -1.524535,
   17.777778,    12.403803,    -5.903803,    -2.240623,
   20.000000,    12.202646,    -5.702646,    -2.943236,
   22.222222,    11.947646,    -5.447646,    -3.628148,
   24.444444,    11.640338,    -5.140338,    -4.291239,
   26.666667,    11.282569,    -4.782569,    -4.928523,
   28.888889,    10.876491,    -4.376491,    -5.536166,
   31.111111,    10.424547,    -3.924547,    -6.110513,
   33.333333,     9.929454,    -3.429454,    -6.648112,
   35.555556,     9.394191,    -2.894191,    -7.

### Exercise 4: Explore plotting in python

For this, check out https://matplotlib.org/3.1.1/gallery/lines_bars_and_markers/simple_plot.html#sphx-glr-gallery-lines-bars-and-markers-simple-plot-py

I copied the demo code below so you can modify it to plot $(\sigma_{x}'/\tau_{xy}')$ and  $(\sigma_{y}'/-\tau_{xy}')$ curves from the data collected in `sigmax_values`, `sigmay_values`, `tauxy_values`

In [ ]:
import matplotlib
import matplotlib.pyplot as plt
import numpy as np

# Data for plotting
t = np.arange(0.0, 2.0, 0.01)
s = 1 + np.sin(2 * np.pi * t)

fig, ax = plt.subplots()
ax.plot(t, s)

ax.set(xlabel='time (s)', ylabel='voltage (mV)',
       title='About as simple as it gets, folks')
ax.grid()

# I comment writing the figure to file ... makes this quicker
# fig.savefig("test.png")
plt.show()

**Remark**: This was just to wet your appetite! We will talk much more about `numpy` and `matplotlib` next week.

## Exercise 5: List versus dictionary

You could express the stress state as

1. three independent variables - the way we did it thus far

2. a list of three values: `stress = `$[\sigma_x, \sigma_y, \tau_{xy}]$ and replace

        sigma_x --> stress[0]
        sigma_y --> stress[2]
        tau_xy  --> stress[3]
        
3. a dictionary 

        stress = {'sigx': 12.0, 'sigy': -5.5, 'tau':3.50}
        
   and replace

        sigma_x --> stress['sigx']
        sigma_y --> stress['sigy']
        tau_xy  --> stress['tau']

**Discuss**:

1. Pros and cons for coding the equations
2. Pros and cons for coding the function(s)
3. How woud code and function change when switching from 2D to 3D (6 components instead of 3 components)

Feel free to implement an alternative form of stressTrannsform() that looks like this:

In [ ]:
def stressTransform(stress, theta):
    # YOUR CODE
    # stress2 is a list or dictionary, just as stress is
    
    return stress2

## Homework questions

We can all learn from your questions!